qiskit/_jobprocessor.py

```diff
@@ -0,0 +1,371 @@
+            if doCompile:
+                for circuit in self.circuits:
+                    formatted_circuits.append(None)
+            else:
+                if backend in self._local_backends:
+                    for circuit in self.circuits:
+                        formatted_circuits.append(openquantumcompiler.dag2json(circuit))
+                else:
+                    for circuit in self.circuits:
+                        formatted_circuits.append(circuit.qasm(qeflag=True))
+            # create circuit component of qobj
+            circuitRecords = []
+            if circuit_configs is None:
+                config = {'coupling_map': None,
+                          'basis_gates': 'u1,u2,u3,cx,id',
+                          'layout': None,
+                          'seed': seed}
+                circuit_configs = [config] * len(self.circuits)
+            for circuit, fcircuit, name, config in zip(self.circuits,
+                                                       formatted_circuits,
+                                                       self.names,
+                                                       circuit_configs):
+                record = {
+                    'name': name,
+                    'compiled_circuit': None if doCompile else fcircuit,
+                    'compiled_circuit_qasm': None if doCompile else fcircuit,
+                    'circuit': circuit,
+                    'config': config
+                }
+                circuitRecords.append(record)
+            qobjid = ''.join([random.choice(
+                string.ascii_letters + string.digits) for i in range(10)])
+            self.qobj = {'id': qobjid,
+                         'config': {
+                             'max_credits': resources['max_credits'],
+                             'shots': shots,
+                             'backend': backend
+                         },
+                         'circuits': circuitRecords
+            }
+        self.seed = seed
+        self.result = None
+        self.doCompile = doCompile
+
+
+class JobProcessor():
+    """
+    process a bunch of jobs and collect the results
+    """
+    def __init__(self, q_jobs, callback, max_workers=1, token=None, url=None, api=None):
+        """
+        Args:
+            q_jobs (QuantumJob | QuantumJob list | qobj): QuantumJob, list of
+                QuantumJob objects, or qobj dictionary.
+            online (bool): If true, establishes a connection to online server. This
+                should be set to True of /any/ of the jobs in q_jobs use a networked
+                backend.
+            max_workers (int): The maximum number of workers to use.
+            token (str): server API token
+            url (str): server URL.
+            api (IBMQuantumExperience): API instance to use. If set,
+                /token/ and /url/ are ignored.
+        """
+        # check whether qobj was supplied.
+        if (isinstance(q_jobs, dict)
```

Comment:
I'm wondering if with the addition of the `QuantumProgram.run_*()` methods and the automatic conversion to lists at `QuantumProgram._run_internal()`, it can be assumed that `q_jobs` will always a `QuantumJob list`?

Reply:
Yeah, good catch! q_jobs should be always a list of QuantumJobs objects, so all this code can be simplified a lot.

Reply:
This is not related to the question, but do you think we need a QuantumJob or should we just put it's attributes in the 'config' key of a qobj? We can maybe put this off until after the merge commit.

Reply:
I really like the abstraction that QuantumJob offers, it makes things easier for us. So I would say that let's keep it for now.